src/util/Taggable.h

```diff
@@ -0,0 +1,101 @@
+#ifndef RIPPLE_UTIL_TAGGABLE_H
+#define RIPPLE_UTIL_TAGGABLE_H
+
+#include <ostream>
+#include <string>
+
+namespace util {
+namespace detail {
+
+/**
+ * @brief This strategy uses an `atomic_uint64_t` to remain lock free.
+ */
+struct UIntTagStrategy
+{
+    static std::string
+    next();
+};
+
+/**
+ * @brief This strategy uses `boost::uuids::uuid` with a static random generator
+ * and a mutex
+ */
+struct UUIDTagStrategy
+{
+    static std::string
+    next();
+};
+
+}  // namespace detail
+
+/**
+ * @brief A base class that allows attaching a tag to a subclass.
+ * @tparam Strategy Specifies how the tag is generated
+ */
+template <typename Strategy = detail::UUIDTagStrategy>
+class Taggable
+{
+private:
+    /**
+     * @brief A simple wrapper for generating and storing tags in a
+     * thread-safe manner.
+     */
+    class Tag final
+    {
+        std::string tag_ = Strategy::next();
+
+    public:
+        /**
+         * @return std::string const& The string representation of the tag
+         */
+        std::string const&
+        str() const
+        {
+            return tag_;
+        }
+
+        /**
+         * @brief Support for ostream (i.e. `std::cout` etc.)
+         *
+         * @param os The stream to stream to
+         * @param tag The tag to stream
+         * @return std::ostream& Stream for chaining
+         */
+        friend std::ostream&
+        operator<<(std::ostream& os, Tag const& tag)
+        {
+            os << tag.str();
+            return os;
+        }
+    };
+
+protected:
+    /**
+     * @brief Only allows subclassing
+     */
+    Taggable() = default;
+
+public:
+    virtual ~Taggable() = default;
+
+    /**
+     * @brief Getter for the session tag
+     * @return Tag const& The ssesion tag
+     */
+    Tag const&
+    tag() const
+    {
+        return tag_;
+    }
+
+private:
+    Tag tag_;
+};
+
+using UUIDTaggable = Taggable<detail::UUIDTagStrategy>;
+using UIntTaggable = Taggable<detail::UIntTagStrategy>;
+using DefaultTaggable = UUIDTaggable;
```

Comment:
Let's make this a configuration parameter, and default to UInt, for performance reasons.

Reply:
I defaulted it to `null` (nop) if nothing is specified in the config. i'll set the example config file to `uint`.